04_basic_struct/module/hello.c

```diff
@@ -0,0 +1,99 @@
+#define pr_fmt(fmt) "%s: " fmt,  KBUILD_MODNAME
+
+#include <linux/module.h> // Core header for loading LKMs into the kernel
+#include <linux/kobject.h>
+#include <linux/err.h>
+#include <linux/string.h>
+#include <linux/sysfs.h>
+#include <linux/list.h>
+#include <linux/slab.h>
+
+struct node {
+	char* value;
+	struct list_head list_member;
+};
+
+static LIST_HEAD(nodeHead);
+
+static void add_node(char* value, struct list_head *head)
+{
+    struct node *nodePtr = (struct node *)kmalloc(sizeof(struct node), GFP_USER);
+    // assert(fooPtr != NULL); TODO use PANIC here
+    
+    nodePtr->value = value;
+    INIT_LIST_HEAD(&nodePtr->list_member);
+    list_add(&(nodePtr->list_member), head);
+}
+
+static void display(struct list_head *head)
+{
+    struct list_head *iter;
+    struct node *nodePtr;
+
+    list_for_each(iter, head) {
+        nodePtr = list_entry(iter, struct node, list_member);
+		printk(KERN_INFO "%s", nodePtr->value);
+    }
+}
+
+static ssize_t list_show(struct kobject *kobj, struct kobj_attribute *attr,
+	char *buf)
+{
+	// struct list_head *iter;
+    // struct node *nodePtr;
+
+	// char outStr[] = "";
+
+    // list_for_each(iter, &nodeHead) {
+        // nodePtr = list_entry(iter, struct node, list_member);
+		// printk(KERN_INFO "Buf size: %ld", strlen(buf));
+		// sprintf(buf, "%s\n", nodePtr->value);
+		// strcat(outStr, nodePtr->value);
+    // }
+    // return sprintf(buf, "%s", outStr);
```

Comment:
`outStr` - is a concatenation of all strings in the list.
For some reason `outStr` cannot be passed to the `buf`.
What is the correct way of returning all list data from the module?